lib/propolis/src/hw/virtio/softnpu.rs

```
@@ -0,0 +1,1042 @@
+
+            let pipeline = pipeline.clone();
+            let uart = uart.clone();
+            let log = log.clone();
+            handle_management_message(msg, pipeline, uart, radix, log.clone());
+            info!(log, "handled management message");
+        }
+    }
+}
+
+impl Entity for SoftNpu {
+    fn type_name(&self) -> &'static str {
+        "softnpu"
+    }
+
+    fn start(&self) {
+        let mut booted = self.booted.lock().unwrap();
+        if *booted {
+            return;
+        }
+        self.run_management_handler_thread();
+        for i in 0..self.pci_port.data_handles.len() {
+            info!(self.log, "starting ingress packet handler for port {}", i);
+
+            PacketHandler::run_ingress_packet_handler_thread(
+                i,
+                self.pci_port.data_handles.clone(),
+                self.pci_port.virtio_state.clone(),
+                self.pci_port.pipeline.clone(),
+                self.log.clone(),
+            );
+        }
+        *booted = true
+    }
+
+    fn migrate(&'_ self) -> Migrator<'_> {
+        Migrator::NonMigratable
+    }
+}
+
+impl PciVirtioSoftNpuPort {
+    pub fn new(
+        mac: [u8; 6],
+        data_handles: Vec<dlpi::DlpiHandle>,
+        virtio: Arc<PortVirtioState>,
+        pipeline: Arc<Mutex<Option<(Library, Box<dyn Pipeline>)>>>,
+        log: Logger,
+    ) -> Arc<Self> {
+        Arc::new(PciVirtioSoftNpuPort {
+            mac,
+            data_handles,
+            pipeline,
+            log,
+            virtio_state: virtio.clone(),
+        })
+    }
+
+    fn handle_guest_virtio_request(&self, vq: &Arc<VirtQueue>) {
+        if vq.id == 0 {
+            return self.handle_q0_req(vq);
+        }
+
+        let mem = match self.virtio_state.pci_state.acc_mem.access() {
+            Some(mem) => mem,
+            None => {
+                warn!(self.log, "failed to access virtio memory");
+                return;
+            }
+        };
```

I would probably just silently return like the `pop_avail` check below.

---

done